File: common/src/main/java/com/wolfyscript/utilities/gui/components/StackInputSlotImpl.java
package com.wolfyscript.utilities.gui.components;

import com.wolfyscript.utilities.KeyedStaticId;
import com.wolfyscript.utilities.WolfyUtils;
import com.wolfyscript.utilities.gui.rendering.PropertyPosition;
import com.wolfyscript.utilities.gui.rendering.RenderProperties;
import com.wolfyscript.utilities.platform.adapters.ItemStack;
import com.wolfyscript.utilities.gui.*;
import com.wolfyscript.utilities.gui.callback.InteractionCallback;
import org.jetbrains.annotations.NotNull;

import java.util.function.Consumer;

@KeyedStaticId(key = "stack_input_slot")
public class StackInputSlotImpl extends AbstractComponentImpl implements Interactable, StackInputSlot {

    private final Consumer<ItemStack> onValueChange;
    private final InteractionCallback interactionCallback;
    private final ItemStack value;

    public StackInputSlotImpl(String internalID, WolfyUtils wolfyUtils, Component parent, Consumer<ItemStack> onValueChange, InteractionCallback interactionCallback, ItemStack value, RenderProperties properties) {
        super(internalID, wolfyUtils, parent, properties);
        this.onValueChange = onValueChange;
        this.interactionCallback = (holder, details) -> {
            // TODO
//            if (details instanceof ClickInteractionDetailsImpl clickInteractionDetails) {
//                InventoryClickEvent event = clickInteractionDetails.getClickEvent();
//                InteractionUtils.applyItemFromInteractionEvent(event.getSlot(), event, Set.of(), itemStack -> {
//                    onValueChange.accept(new ItemStackImpl((WolfyUtilsBukkit) holder.getViewManager().getWolfyUtils(), itemStack));
//                });
//            }
            return interactionCallback.interact(holder, details);
        };
        this.value = value;
    }

    @Override
    public int width() {
        return 1;
    }

    @Override
    public int height() {
        return 1;
    }

    @Override
    public InteractionResult interact(GuiHolder guiHolder, InteractionDetails interactionDetails) {
        return interactionCallback.interact(guiHolder, interactionDetails);
    }

    @Override
    public InteractionCallback interactCallback() {
        return interactionCallback;
    }

    @Override
    public ItemStack value() {
        return value;
    }

    @Override
    public void remove(@NotNull ViewRuntimeImpl viewRuntimeImpl, long nodeId, long parentNode) {
        viewRuntimeImpl.getRenderingGraph().removeNode(nodeId);
    }

    @Override
    public void insert(@NotNull ViewRuntimeImpl viewRuntimeImpl, long parentNode) {
        long id = viewRuntimeImpl.getRenderingGraph().addNode(this);
        viewRuntimeImpl.getRenderingGraph().insertNodeChild(id, parentNode);
    }
}

File: common/src/main/java/com/wolfyscript/utilities/gui/components/StackInputSlotImpl.kt
package com.wolfyscript.utilities.gui.components

import com.fasterxml.jackson.annotation.JacksonInject
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.google.inject.Inject
import com.wolfyscript.utilities.KeyedStaticId
import com.wolfyscript.utilities.WolfyUtils
import com.wolfyscript.utilities.functions.ReceiverBiFunction
import com.wolfyscript.utilities.functions.ReceiverConsumer
import com.wolfyscript.utilities.gui.BuildContext
import com.wolfyscript.utilities.gui.ViewRuntime
import com.wolfyscript.utilities.gui.ViewRuntimeImpl
import com.wolfyscript.utilities.gui.interaction.ClickInteractionDetails
import com.wolfyscript.utilities.gui.interaction.ClickType
import com.wolfyscript.utilities.platform.adapters.ItemStack
import java.util.function.Consumer
import javax.annotation.Nullable

@ComponentImplementation(base = StackInputSlot::class)
@KeyedStaticId(key = "stack_input_slot")
class StackInputSlotImpl @JsonCreator @Inject constructor(
    @JsonProperty("id") id: String,
    @JacksonInject("wolfyUtils") wolfyUtils: WolfyUtils,
    @JacksonInject("context") private val context: BuildContext,
    @Nullable @JacksonInject("parent") parent: Component? = null,
) : AbstractComponentImpl<StackInputSlot>(
    id, wolfyUtils, parent
), StackInputSlot {

    override var onValueChange: Consumer<ItemStack?>? = null
    override var onClick: ReceiverConsumer<ClickInteractionDetails>? = null
    override var acceptStack: ReceiverBiFunction<ClickType, ItemStack, Boolean>? = null
    override var canPickUpStack: ReceiverBiFunction<ClickType, ItemStack, Boolean>? = null
    override var value: ItemStack? = null

    override fun width(): Int {
        return 1
    }

    override fun height(): Int {
        return 1
    }

    override fun remove(runtime: ViewRuntime, nodeId: Long, parentNode: Long) {
        (runtime as ViewRuntimeImpl).renderingGraph.removeNode(nodeId)
    }

    override fun insert(runtime: ViewRuntime, parentNode: Long) {
        runtime as ViewRuntimeImpl
        val id = runtime.renderingGraph.addNode(this)
        runtime.renderingGraph.insertNodeChild(id, parentNode)
    }
}
